frontend/src/components/Header/StopClock.js

```diff
@@ -0,0 +1,98 @@
+// @flow
+import React, { Component } from 'react';
+import CircularProgressbar from 'react-circular-progressbar';
+import 'react-circular-progressbar/dist/styles.css';
+
+type Props = {
+  startTime: number, // Date.now() ms at start
+  countdown: number, // delta in seconds
+  currentTime: number, // Date.now() at present
+  history: *,
+};
+
+type State = {
+  currentTime: number,
+};
+
+type CircularProgressbarProps = {
+  text: string,
+  percentage: number,
+};
+
+const StyledCircularProgressbar = (props: CircularProgressbarProps) => (
+  <CircularProgressbar
+    percentage={props.percentage}
+    text={props.text}
+    strokeWidth={10}
+    styles={{
+      root: {},
+      path: {
+        stroke: '#1c8bf4',
+        strokeLinecap: 'butt',
+        transition: 'stroke-dashoffset 0.5s ease 0s',
+      },
+      trail: {
+        stroke: '#e6e6e6',
+      },
+      text: {
+        fill: '#1c8bf4',
+        fontSize: '30px',
+      },
+    }}
+  />
+);
+
+const getPercRemaining = (
+  currentTime: number,
+  startTime: number,
+  countdown: number
+) => {
+  return (currentTime - startTime) / (countdown * 1000);
+};
+
+const getTimeLabel = (secondsRemaining: number) =>
+  secondsRemaining < 60
+    ? `${Math.ceil(secondsRemaining)}s`
+    : `${Math.ceil(secondsRemaining / 60)}m`;
+
+class StopClock extends Component<Props, State> {
+  state: State = {
+    currentTime: Date.now(),
+  };
+
+  tick = () => {
+    this.setState({
+      currentTime: Date.now(),
+    });
+  };
+
+  timer = null;
+  endTime = 0;
+  componentDidMount() {
+    this.timer = setInterval(this.tick, 500);
+    this.endTime = this.props.startTime + 1000 * this.props.countdown;
+  }
+
+  componentWillUnmount() {
+    clearInterval(this.timer);
+  }
+
+  render() {
+    const { currentTime } = this.state;
+    console.log(this.props);
+    console.log(this.state);
+    if (this.endTime != 0 && currentTime >= this.endTime) {
+      this.props.history.push('/welcome');
```

Comment:
You can use the `Redirect` component from react-router here: https://reacttraining.com/react-router/core/api/Redirect

Reply:
Isn't that the equivalent of rendering a DOM redirect rather than just redirecting programmatically? I could also use history.replace to redirect

Reply:
Not quite. `render` can be called any number of times (or multiple times) and so each time its called `this.props.history.push` would add another page to the stack. The `Redirect` component does the `push` in it's `componentDidMount` method so it's sure to only happen once: https://github.com/ReactTraining/react-router/blob/master/packages/react-router/modules/Redirect.js#L32